mount/upgrader.go

```diff
@@ -0,0 +1,160 @@
+package mount
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+// Upgrader serves as a bridge to upgrade a Mount into one with full-featured
+// Reader capabilities. It does this by caching a transient copy as file if
+// the original mount type does not support all access patterns.
+//
+// If the underlying mount is already fully-featured, the Upgrader is
+// acts as a noop.
+//
+// TODO perform refcounts so we can track inactive transient files.
+// TODO provide root directory for temp files (or better: temp file factory function).
+type Upgrader struct {
+	underlying  Mount
+	passthrough bool
+
+	lk        sync.Mutex
+	transient string
+	// TODO refs int
+}
+
+var _ Mount = (*Upgrader)(nil)
+
+// Upgrade constructs a new Upgrader for the underlying Mount.
+func Upgrade(underlying Mount, initial string) (*Upgrader, error) {
+	ret := &Upgrader{underlying: underlying}
+
+	info := underlying.Info()
+	if !info.AccessSequential {
+		return nil, fmt.Errorf("underlying mount must support sequential access")
+	}
+	if info.AccessSequential && info.AccessSeek && info.AccessRandom {
+		ret.passthrough = true
+		return ret, nil
+	}
+
+	if initial != "" {
+		if _, err := os.Stat(initial); err == nil {
+			ret.transient = initial
+			return ret, nil
+		}
+	}
+
+	return ret, nil
+}
+
+func (u *Upgrader) Fetch(ctx context.Context) (Reader, error) {
+	if u.passthrough {
+		return u.underlying.Fetch(ctx)
+	}
+
+	// determine if the transient is still alive.
+	u.lk.Lock()
+	defer u.lk.Unlock()
+
+	if u.transient != "" {
+		if _, err := os.Stat(u.transient); err == nil {
+			return os.Open(u.transient)
+		}
+	}
+
+	// transient appears to be dead, refetch.
+	if err := u.refetch(ctx); err != nil {
+		return nil, err
+	}
+
+	return os.Open(u.transient)
+}
+
+func (u *Upgrader) Info() Info {
+	return Info{
+		Kind:             KindLocal,
+		URL:              u.underlying.Info().URL,
+		AccessSequential: true,
+		AccessSeek:       true,
+		AccessRandom:     true,
+	}
+}
+
+func (u *Upgrader) Stat(ctx context.Context) (Stat, error) {
+	if u.transient != "" {
+		if stat, err := os.Stat(u.transient); err == nil {
+			ret := Stat{Exists: true, Size: stat.Size()}
+			return ret, nil
+		}
+	}
+	return u.underlying.Stat(ctx)
+}
+
+func (u *Upgrader) Close() error {
+	panic("implement me")
+}
+
+func (u *Upgrader) refetch(ctx context.Context) error {
+	if u.transient != "" {
+		_ = os.Remove(u.transient)
+	}
+	file, err := os.CreateTemp("dagstore", "transient")
```

Comment:
nit: should create the temp after the below `Mount.Stat` piece of code is successful.

Reply:
This needs to change anyway; we shouldn't be using temp files: #26.